packages/composite-checkout-wpcom/src/hooks/use-shopping-cart.ts

```diff
@@ -169,45 +169,53 @@ export function useShoppingCart(
 			} );
 		}
 
-		return () => ( isSubscribed = false );
+		return () => {
+			isSubscribed = false;
+		};
 	}, [ getServerCart, cacheStatus ] );
 
 	// Asynchronously re-validate when the cache is dirty.
 	useEffect( () => {
+		let isSubscribed = true;
+
 		const fetchAndUpdate = async () => {
 			debug( 'sending cart with responseCart', responseCart );
 			const response = await setServerCart( prepareRequestCart( responseCart ) );
 			debug( 'cart sent; new responseCart is', response );
-			setResponseCart( response );
+			isSubscribed && setResponseCart( response );
 
 			if ( couponStatus === 'pending' ) {
 				if ( response.is_coupon_applied ) {
 					showAddCouponSuccessMessage( response.coupon );
-					setCouponStatus( 'applied' );
+					isSubscribed && setCouponStatus( 'applied' );
 				}
 
 				if ( ! response.is_coupon_applied && response.coupon_discounts_integer?.length <= 0 ) {
-					setCouponStatus( 'invalid' );
+					isSubscribed && setCouponStatus( 'invalid' );
 				}
 
 				if ( ! response.is_coupon_applied && response.coupon_discounts_integer?.length > 0 ) {
-					setCouponStatus( 'rejected' );
+					isSubscribed && setCouponStatus( 'rejected' );
 				}
 			}
 
-			setCacheStatus( 'valid' );
+			isSubscribed && setCacheStatus( 'valid' );
 		};
 
 		debug( 'considering sending cart to server; cacheStatus is', cacheStatus );
 		if ( cacheStatus === 'invalid' ) {
 			debug( 'updating the cart' );
-			setCacheStatus( 'pending' );
+			isSubscribed && setCacheStatus( 'pending' );
 			fetchAndUpdate().catch( error => {
 				// TODO: figure out what to do here
-				setCacheStatus( 'error' );
+				isSubscribed && setCacheStatus( 'error' );
 				debug( 'error while fetching cart', error );
 			} );
 		}
+
+		return () => {
+			isSubscribed = false;
```

Comment:
This may cause some unexpected behavior that I didn't predict when I suggested it.

When your `useEffect` starts, depending on the cache status, it may call `fetchAndUpdate()` and either way it then exits (not waiting for that function to complete). Let's ignore the async function for the moment and follow the rest of the code.

Due to the nature of `useEffect`, it will then ignore all that on subsequent renders unless either the component is unmounted or one of the dependencies (`cacheStatus, couponStatus, responseCart, showAddCouponSuccessMessage`) changes.

If one of them changes, and we'll come back to which ones are likely to change, it will first set `isSubscribed` to false _for the previous scope of `useEffect`_ (not the scope it's about to create), and call the `useEffect` again. This time it will do the same check on the cache status and may call `fetchAndUpdate()` again; all of this regardless of what happened in the async call the first time.

So what would happen if the first call happens, then we wait a few seconds before `fetchAndUpdate()` gets its first response from the server, but before it does, one of the dependencies changes? The "paused" `fetchAndUpdate()` function which is waiting for data would be told (via `isSubscribed`) that it should essentially cancel any further work. This all happens before the second call of `useEffect` begins. Now it would start the call, and perhaps the cache status is such that we _do not_ want to call `fetchAndUpdate()` again, so the `useEffect` exits, and has done nothing.

so you're kind of in a race condition with the `cacheStatus`. depending on _why_ the dependencies changed and which one, you might _not_ want to cancel a running operation. Another way to think about it is that maybe the cache status shouldn't be the only thing that keeps the async callback from being restarted. Or perhaps that there's dependencies you should be ignoring. There's a lot of ways to look at the issue.

Reply:
After thinking about this some more, I think a better solution to address this problem, as well as a couple of others, is to bite the bullet and refactor this hook to use the reducer pattern. I just rebased this with all changes except the `isSubscribed` pattern, so it's currently in the "working by accident" state. The refactoring is a significant enough change that I think it belongs in a separate PR. I'll try to get that up asap.

Reply:
> belongs in a separate PR

Here it is: #39356